File: provable-sdk-js/src/lightnet.ts
/**
 * Lightnet API client - Database, Hash, and Merkle operations
 */

import { API_ROUTES, getKayrosUrl, getDefaultHeaders } from './config';
import type {
  APIResponse,
  DatabaseQuery,
  HashRecord,
  DatabaseStats,
  ColumnInfo,
  TableBrowseRequest,
  DatabaseRecord,
  HashVerifyRequest,
  HashVerifyResult,
  ComputeHashRequest,
  ComputeHashResponse,
  SingleHashRequest,
  SingleHashResponse,
  GetRecordByDataItemResponse,
  MerkleProofResponse,
  HashExistenceRequest,
  HashExistenceResponse,
  HashBatchRequest,
  HashBatchResponse,
} from './types';
import type { ApiKeyOptions } from './options';

// Database Operations

/**
 * Query hash records from the database
 * @param query - Database query parameters
 * @returns Promise with hash records
 */
export async function query_hashes(query: DatabaseQuery): Promise<APIResponse<HashRecord[]>> {
  const url = getKayrosUrl('/api/database/query');

  const response = await fetch(url, {
    method: 'POST',
    headers: getDefaultHeaders(),
    body: JSON.stringify(query),
  });

  if (!response.ok) {
    throw new Error(`API error: ${response.status} ${response.statusText}`);
  }

  return await response.json() as APIResponse<HashRecord[]>;
}

/**
 * Get database statistics
 * @returns Promise with database stats
 */
export async function get_database_stats(): Promise<APIResponse<DatabaseStats>> {
  const url = getKayrosUrl('/api/database/stats');

  const response = await fetch(url, {
    method: 'GET',
    headers: getDefaultHeaders(),
  });

  if (!response.ok) {
    throw new Error(`API error: ${response.status} ${response.statusText}`);
  }

  return await response.json() as APIResponse<DatabaseStats>;
}

/**
 * Get the most recent hash records
 * @param limit - Number of records to retrieve (default 50)
 * @returns Promise with latest hash records
 */
export async function get_latest_hashes(limit: number = 50): Promise<APIResponse<HashRecord[]>> {
  const url = getKayrosUrl(`/api/database/latest?limit=${limit}`);

  const response = await fetch(url, {
    method: 'GET',
    headers: getDefaultHeaders(),
  });

  if (!response.ok) {
    throw new Error(`API error: ${response.status} ${response.statusText}`);
  }

  return await response.json() as APIResponse<HashRecord[]>;
}

/**
 * Get all database tables
 * @returns Promise with table names
 */
export async function get_tables(): Promise<APIResponse<string[]>> {
  const url = getKayrosUrl('/api/database/tables');

  const response = await fetch(url, {
    method: 'GET',
    headers: getDefaultHeaders(),
  });

  if (!response.ok) {
    throw new Error(`API error: ${response.status} ${response.statusText}`);
  }

  return await response.json() as APIResponse<string[]>;
}

/**
 * Get schema for a specific table
 * @param tableName - Name of the table
 * @returns Promise with column information
 */
export async function get_table_schema(tableName: string): Promise<APIResponse<ColumnInfo[]>> {
  const url = getKayrosUrl(`/api/database/schema?table=${encodeURIComponent(tableName)}`);

  const response = await fetch(url, {
    method: 'GET',
    headers: getDefaultHeaders(),
  });

  if (!response.ok) {
    throw new Error(`API error: ${response.status} ${response.statusText}`);
  }

  return await response.json() as APIResponse<ColumnInfo[]>;
}

/**
 * Browse table data with pagination
 * @param request - Table browse parameters
 * @returns Promise with table rows
 */
export async function browse_table(request: TableBrowseRequest): Promise<APIResponse<unknown[]>> {
  const url = getKayrosUrl('/api/database/browse');

  const response = await fetch(url, {
    method: 'POST',
    headers: getDefaultHeaders(),
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    throw new Error(`API error: ${response.status} ${response.statusText}`);
  }

  return await response.json() as APIResponse<unknown[]>;
}

/**
 * Get a record by UUID
 * @param uuid - Record UUID (hex string)
 * @returns Promise with database record
 */
export async function get_record(uuid: string): Promise<APIResponse<DatabaseRecord>> {
  const url = getKayrosUrl(`/api/database/record?uuid=${encodeURIComponent(uuid)}`);

  const response = await fetch(url, {
    method: 'GET',
    headers: getDefaultHeaders(),
  });

  if (!response.ok) {
    throw new Error(`API error: ${response.status} ${response.statusText}`);
  }

  return await response.json() as APIResponse<DatabaseRecord>;
}

/**
 * Get a record by UUID with previous hash
 * @param uuid - Record UUID (hex string)
 * @returns Promise with database record including prev_hash
 */
export async function get_record_with_prev_hash(uuid: string): Promise<APIResponse<DatabaseRecord>> {
  const url = getKayrosUrl(`/api/database/record-with-prev?uuid=${encodeURIComponent(uuid)}`);

  const response = await fetch(url, {
    method: 'GET',
    headers: getDefaultHeaders(),
  });

  if (!response.ok) {
    throw new Error(`API error: ${response.status} ${response.statusText}`);
  }

  return await response.json() as APIResponse<DatabaseRecord>;
}

/**
 * Get records by data type and data item.
 */
export async function get_record_by_data_item(
  dataType: string,
  dataItem: string,
  options?: ApiKeyOptions & { limit?: number }
): Promise<GetRecordByDataItemResponse> {
  let query = `${API_ROUTES.GET_RECORD_BY_DATA_ITEM}?data_type=${encodeURIComponent(dataType)}&data_item=${encodeURIComponent(dataItem)}`;
  if (typeof options?.limit === 'number' && options.limit > 0) {
    query += `&limit=${options.limit}`;
  }

  const response = await fetch(getKayrosUrl(query), {
    method: 'GET',
    headers: getDefaultHeaders(options),
  });

  if (!response.ok) {
    throw new Error(`API error: ${response.status} ${response.statusText}`);
  }

  return await response.json() as GetRecordByDataItemResponse;
}

// Hash Operations

/**
 * Verify a hash computation
 * @param request - Hash verification request
 * @returns Promise with hash verification result
 */
export async function verify_hash(request: HashVerifyRequest): Promise<APIResponse<HashVerifyResult>> {
  const url = getKayrosUrl('/api/verify-hash');

  const response = await fetch(url, {
    method: 'POST',
    headers: getDefaultHeaders(),
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    throw new Error(`API error: ${response.status} ${response.statusText}`);
  }

  return await response.json() as APIResponse<HashVerifyResult>;
}

/**
 * Compute hash from hex input
 * @param request - Compute hash request
 * @returns Promise with computed hash result
 */
export async function compute_hash_from_hex(
  request: ComputeHashRequest,
  options?: ApiKeyOptions
): Promise<ComputeHashResponse> {
  const url = getKayrosUrl(API_ROUTES.COMPUTE_HASH_FROM_HEX);

  const response = await fetch(url, {
    method: 'POST',
    headers: getDefaultHeaders(options),
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    throw new Error(`API error: ${response.status} ${response.statusText}`);
  }

  return await response.json() as ComputeHashResponse;
}

// gRPC Operations

/**
 * Send a single gRPC request to Lightnet
 * @param request - Single hash request with data_type and data_item
 * @returns Promise with gRPC response
 */
export async function send_single_grpc_request(request: SingleHashRequest): Promise<APIResponse<SingleHashResponse>> {
  const url = getKayrosUrl(API_ROUTES.PROVE_SINGLE_HASH);

  const response = await fetch(url, {
    method: 'POST',
    headers: getDefaultHeaders(),
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    throw new Error(`API error: ${response.status} ${response.statusText}`);
  }

  return await response.json() as APIResponse<SingleHashResponse>;
}

// Merkle Proof Operations

/**
 * Generate a Merkle proof for a specific hash
 * @param request - Merkle proof generation request
 * @returns Promise with Merkle proof
 */
export async function get_merkle_proof(
  request: { data_type: string; hash?: string; position?: number },
  options?: ApiKeyOptions
): Promise<MerkleProofResponse> {
  const params = new URLSearchParams();
  params.set('data_type', request.data_type);
  if (request.hash) {
    params.set('hash', request.hash);
  }
  if (typeof request.position === 'number') {
    params.set('position', String(request.position));
  }

  const response = await fetch(getKayrosUrl(`${API_ROUTES.GET_MERKLE_PROOF}?${params.toString()}`), {
    method: 'GET',
    headers: getDefaultHeaders(options),
  });

  if (!response.ok) {
    throw new Error(`API error: ${response.status} ${response.statusText}`);
  }

  return await response.json() as MerkleProofResponse;
}

export async function verify_hash_existence(
  request: HashExistenceRequest,
  options?: ApiKeyOptions
): Promise<HashExistenceResponse> {
  const response = await fetch(getKayrosUrl(API_ROUTES.VERIFY_HASH_EXISTENCE), {
    method: 'POST',
    headers: getDefaultHeaders(options),
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    throw new Error(`API error: ${response.status} ${response.statusText}`);
  }

  return await response.json() as HashExistenceResponse;
}

export async function verify_hash_batch(
  request: HashBatchRequest,
  options?: ApiKeyOptions
): Promise<HashBatchResponse> {
  const response = await fetch(getKayrosUrl(API_ROUTES.VERIFY_HASH_BATCH), {
    method: 'POST',
    headers: getDefaultHeaders(options),
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    throw new Error(`API error: ${response.status} ${response.statusText}`);
  }

  return await response.json() as HashBatchResponse;
}

File: provable-sdk-py/provable_sdk/lightnet.py
"""
Lightnet API client - Database, Hash, and Merkle operations
"""

import requests
from typing import Any, Dict, List
from urllib.parse import urlencode

from .config import get_kayros_url, API_ROUTES, get_default_headers
from .types import (
    APIResponse,
    DatabaseQuery,
    HashRecord,
    DatabaseStats,
    ColumnInfo,
    TableBrowseRequest,
    DatabaseRecord,
    HashVerifyRequest,
    HashVerifyResult,
    ComputeHashRequest,
    ComputeHashResponse,
    SingleHashRequest,
    SingleHashResponse,
    GetRecordByDataItemResponse,
    MerkleProofResponse,
    HashExistenceRequest,
    HashExistenceResponse,
    HashBatchRequest,
    HashBatchResponse,
)


# Database Operations

def query_hashes(query: DatabaseQuery) -> APIResponse:
    """
    Query hash records from the database

    Args:
        query: Database query parameters

    Returns:
        API response with hash records

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = get_kayros_url('/api/database/query')
    response = requests.post(url, json=query, headers=get_default_headers())
    response.raise_for_status()
    return response.json()


def get_database_stats() -> APIResponse:
    """
    Get database statistics

    Returns:
        API response with database stats

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = get_kayros_url('/api/database/stats')
    response = requests.get(url, headers=get_default_headers())
    response.raise_for_status()
    return response.json()


def get_latest_hashes(limit: int = 50) -> APIResponse:
    """
    Get the most recent hash records

    Args:
        limit: Number of records to retrieve (default 50)

    Returns:
        API response with latest hash records

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = get_kayros_url(f'/api/database/latest?limit={limit}')
    response = requests.get(url, headers=get_default_headers())
    response.raise_for_status()
    return response.json()


def get_tables() -> APIResponse:
    """
    Get all database tables

    Returns:
        API response with table names

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = get_kayros_url('/api/database/tables')
    response = requests.get(url, headers=get_default_headers())
    response.raise_for_status()
    return response.json()


def get_table_schema(table_name: str) -> APIResponse:
    """
    Get schema for a specific table

    Args:
        table_name: Name of the table

    Returns:
        API response with column information

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = get_kayros_url(f'/api/database/schema?table={requests.utils.quote(table_name)}')
    response = requests.get(url, headers=get_default_headers())
    response.raise_for_status()
    return response.json()


def browse_table(request: TableBrowseRequest) -> APIResponse:
    """
    Browse table data with pagination

    Args:
        request: Table browse parameters

    Returns:
        API response with table rows

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = get_kayros_url('/api/database/browse')
    response = requests.post(url, json=request, headers=get_default_headers())
    response.raise_for_status()
    return response.json()


def get_record(uuid: str) -> APIResponse:
    """
    Get a record by UUID

    Args:
        uuid: Record UUID (hex string)

    Returns:
        API response with database record

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = get_kayros_url(f'/api/database/record?uuid={requests.utils.quote(uuid)}')
    response = requests.get(url, headers=get_default_headers())
    response.raise_for_status()
    return response.json()


def get_record_with_prev_hash(uuid: str) -> APIResponse:
    """
    Get a record by UUID with previous hash

    Args:
        uuid: Record UUID (hex string)

    Returns:
        API response with database record including prev_hash

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = get_kayros_url(f'/api/database/record-with-prev?uuid={requests.utils.quote(uuid)}')
    response = requests.get(url, headers=get_default_headers())
    response.raise_for_status()
    return response.json()


def get_record_by_data_item(data_type: str, data_item: str, api_key: str = None, limit: int = None) -> GetRecordByDataItemResponse:
    """
    Get records by data_type and data_item.
    """
    params = {
        "data_type": data_type,
        "data_item": data_item,
    }
    if limit is not None and limit > 0:
        params["limit"] = limit

    url = get_kayros_url(f"{API_ROUTES['GET_RECORD_BY_DATA_ITEM']}?{urlencode(params)}")
    response = requests.get(url, headers=get_default_headers(api_key))
    response.raise_for_status()
    return response.json()


# Hash Operations

def verify_hash(request: HashVerifyRequest) -> APIResponse:
    """
    Verify a hash computation

    Args:
        request: Hash verification request

    Returns:
        API response with hash verification result

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = get_kayros_url('/api/verify-hash')
    response = requests.post(url, json=request, headers=get_default_headers())
    response.raise_for_status()
    return response.json()


def compute_hash_from_hex(request: ComputeHashRequest, api_key: str = None) -> ComputeHashResponse:
    """
    Compute hash from hex input

    Args:
        request: Compute hash request

    Returns:
        API response with computed hash result

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = get_kayros_url(API_ROUTES["COMPUTE_HASH_FROM_HEX"])
    response = requests.post(url, json=request, headers=get_default_headers(api_key))
    response.raise_for_status()
    return response.json()


# gRPC Operations

def send_single_grpc_request(request: SingleHashRequest) -> APIResponse:
    """
    Send a single gRPC request to Lightnet

    Args:
        request: Single hash request with data_type and data_item

    Returns:
        API response with gRPC response

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = get_kayros_url(API_ROUTES["PROVE_SINGLE_HASH"])
    response = requests.post(url, json=request, headers=get_default_headers())
    response.raise_for_status()
    return response.json()


# Merkle Proof Operations

def get_merkle_proof(data_type: str, *, hash: str = None, position: int = None, api_key: str = None) -> MerkleProofResponse:
    """
    Get a Merkle proof by hash or position.
    """
    params: Dict[str, Any] = {"data_type": data_type}
    if hash is not None:
        params["hash"] = hash
    if position is not None:
        params["position"] = position

    url = get_kayros_url(f"{API_ROUTES['GET_MERKLE_PROOF']}?{urlencode(params)}")
    response = requests.get(url, headers=get_default_headers(api_key))
    response.raise_for_status()
    return response.json()


def verify_hash_existence(request: HashExistenceRequest, api_key: str = None) -> HashExistenceResponse:
    """
    Verify that a hash exists at a specific level and position.
    """
    url = get_kayros_url(API_ROUTES["VERIFY_HASH_EXISTENCE"])
    response = requests.post(url, json=request, headers=get_default_headers(api_key))
    response.raise_for_status()
    return response.json()


def verify_hash_batch(request: HashBatchRequest, api_key: str = None) -> HashBatchResponse:
    """
    Verify multiple hashes at once for a level/start window.
    """
    url = get_kayros_url(API_ROUTES["VERIFY_HASH_BATCH"])
    response = requests.post(url, json=request, headers=get_default_headers(api_key))
    response.raise_for_status()
    return response.json()
